onnxruntime/contrib_ops/cuda/collective/ipc_utils.cc

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "ipc_utils.h"
+#include "mpi_include.h"
+
+namespace onnxruntime {
+namespace cuda {
+namespace collective {
+
+#if defined(USE_MPI) || defined(USE_NCCL)
+
+using namespace onnxruntime;
+
+IpcMemory::IpcMemory(int rank, int world_size, std::size_t buffer_size)
+    : rank_(rank), world_size_(world_size), m_comm_ptrs_(world_size), mbuffer_size_(buffer_size) {
+  ORT_ENFORCE(AllocateIpcMemory() == Status::OK());
+}
+
+Status IpcMemory::AllocateIpcMemory() {
+  CUDA_RETURN_IF_ERROR(cudaMalloc(&m_buffer_ptr_, mbuffer_size_));
+  CUDA_RETURN_IF_ERROR(cudaMemset(m_buffer_ptr_, 0, mbuffer_size_));
```

Comment:
>CUDA_RETURN_IF_ERROR(cudaMemset(m_buffer_ptr_, 0, mbuffer_size_)); [](http://example.com/codeflow?start=2&length=66)

What happens if any of the functions after cudaMalloc errors out? It looks like a memory leak.